src/Vimeo/Vimeo.php

```diff
@@ -546,4 +578,94 @@ private function perform_upload($file_path, $ticket)
+        $file = fopen($file_path, 'r');
+
+        $curl_opts = array(
+            CURLOPT_POST => true,
+            CURLOPT_CUSTOMREQUEST => 'PATCH',
+            CURLOPT_INFILE => $file,
+            CURLOPT_INFILESIZE => filesize($file_path),
+            CURLOPT_UPLOAD => true,
+            CURLOPT_HTTPHEADER => array(
+                'Expect: ',
+                'Content-Type: application/offset+octet-stream',
+                'Tus-Resumable: 1.0.0',
+                'Upload-Offset: {will be replaced}',
+            )
+        );
+
+        // Perform the upload by sending as much to the server as possible and ending when we reach the file size on
+        // the server.
+        $failures = 0;
+        $server_at = 0;
+        do {
+            // The last HTTP header we set has to be `Upload-Offset`, since for resumable uploading to work properly,
+            // we'll need to alter the content of the header for each upload segment request.
+            array_pop($curl_opts[CURLOPT_HTTPHEADER]);
+            $curl_opts[CURLOPT_HTTPHEADER][] = 'Upload-Offset: ' . $server_at;
+
+            fseek($file, $server_at);
+
+            try {
+                $response = $this->_request($url, $curl_opts);
+
+                // Successful upload, so reset the failure counter.
+                $failures = 0;
+
+                if ($response['status'] === 204) {
+                    // If the `Upload-Offset` returned is equal to the size of the video we want to upload, then we've
+                    // fully uploaded the video. If not, continue uploading.
+                    if ($response['headers']['Upload-Offset'] == $file_size) {
+                        break;
+                    }
+
+                    $server_at = $response['headers']['Upload-Offset'];
+                    continue;
+                }
+
+                // If we didn't receive a 204 response from the tus server, then we should verify what's going on before
+                // proceeding to upload more pieces.
+                $verify_response = $this->request($url, array(), 'HEAD');
+                if ($verify_response['status'] !== 200) {
+                    $verify_error = !empty($ticket['body']) ? ' [' . $ticket['body'] . ']' : '';
+                    throw new VimeoUploadException('Unable to verify upload' . $verify_error);
+                }
+
+                if ($verify_response['headers']['Upload-Offset'] == $file_size) {
+                    break;
+                }
+
+                $server_at = $verify_response['headers']['Upload-Offset'];
+            } catch (VimeoRequestException $exception) {
+                // We likely experienced a timeout, but if we experience three in a row, then we should back off and
+                // fail so as to not overwhelm servers that are, probably, down.
+                if ($failures >= 3) {
+                    throw $exception;
+                }
+
+                $failures++;
```

Comment:
Should there be some form of sleep here based on the failure count, like 

```
sleep(4 ^ $failures); // sleep 4, 16, 64 seconds (based on timeouts)
```

Reply:
Implemented.